src/main.cpp

```diff
@@ -95,888 +83,119 @@ DEFINE_bool(
+      console::EchoFailure("Password not submitted",
                   "Requested password for username " + FLAGS_username);
+      cleanup_resources();
       return 1;
     }
-    SetStdinEcho(true);
+    console::SetStdinEcho(true);
   }
 
-  using_history();
-  int history_len = 0;
-  rl_attempted_completion_function = Completer;
   fs::path history_dir = FLAGS_history;
   if (FLAGS_history ==
-      (kDefaultHistoryBaseDir + "/" + kDefaultHistoryMemgraphDir)) {
+      (constants::kDefaultHistoryBaseDir + "/" + constants::kDefaultHistoryMemgraphDir)) {
     // Fetch home dir for user.
-    struct passwd *pw = getpwuid(getuid());
-    history_dir = fs::path(pw->pw_dir) / kDefaultHistoryMemgraphDir;
+    history_dir =
+        utils::GetUserHomeDir() / constants::kDefaultHistoryMemgraphDir;
   }
   if (!utils::EnsureDir(history_dir)) {
-    EchoFailure("History directory doesn't exist", history_dir);
+    console::EchoFailure("History directory doesn't exist",
+                         history_dir.string());
     // Should program exit here or just continue with warning message?
+    cleanup_resources();
     return 1;
   }
-  fs::path history_file = history_dir / kHistoryFilename;
+  fs::path history_file = history_dir / constants::kHistoryFilename;
   // Read history file.
   if (fs::exists(history_file)) {
-    auto ret = read_history(history_file.string().c_str());
+    auto ret =
+        replxx_history_load(replxx_instance, history_file.string().c_str());
     if (ret != 0) {
-      EchoFailure("Unable to read history file", history_file);
+      console::EchoFailure("Unable to read history file",
+                           history_file.string());
       // Should program exit here or just continue with warning message?
+      cleanup_resources();
       return 1;
     }
-    history_len = history_length;
   }
 
   // Save history function. Used to save readline history after each query.
-  auto save_history = [&history_len, history_file] {
+  auto save_history = [history_file, replxx_instance, cleanup_resources] {
     if (!FLAGS_no_history) {
-      int ret = 0;
       // If there was no history, create history file.
       // Otherwise, append to existing history.
-      if (history_len == 0) {
-        ret = write_history(history_file.string().c_str());
-      } else {
-        ret = append_history(1, history_file.string().c_str());
-      }
+      auto ret =
+          replxx_history_save(replxx_instance, history_file.string().c_str());
       if (ret != 0) {
-        EchoFailure("Unable to save history to file", history_file);
+        console::EchoFailure("Unable to save history to file",
+                             history_file.string());
+        cleanup_resources();
         return 1;
       }
-      ++history_len;
     }
     return 0;
   };
 
+#ifdef _WIN32
+// ToDo(the-joksim):
+//  - How to handle shutdown inside a shutdown on Windows? (Windows uses messages
+//  instead of signals.)
```

Comment:
It seems to me it's something that happens once in a blue moon. What's a bit mysterious is that the replxx input loop handles `CTRL+C`, which would cause the console process to send `SIGINT` to `mgconsole` (readline used to do the same), and replxx gracefully handles it by returning a nullptr as `line`, so the application using replxx can decide what to do (it returns from `main` with `1`). `SIGTERM` is rarely sent, and it usually happens by either issuing the `kill` command explicitly, or by some system daemon/service on system shutdown.

What the Unix code does is just prevent handling a second signal that's sent to the mgconsole process, and issuing `std::quick_exit`. Now, `quick_exit` supposedly doesn't clean the resources completely (as opposed to `exit` or the like, I guess).

Having vomited all of the above, I still can't think of a convincing argument why this is necessary, and what would provoke the double shutdown. I guess we want to prevent undefined behavior, whatever that may be. The process hanging?

Reply:
In any case, I have to do a bit more digging to see what's the counterpart on Windows, and how this can be handled. But, honestly, I don't think I've ever encountered this behavior with mgconsole.

Reply:
Then maybe just adding the condensed version of this comment and removing the TODO might be fine.